### PR TITLE
[CodeCamp #122] Support KD algorithm MGD for detection.

### PR DESCRIPTION
**Pull request by TinyTigerPan**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support KD algorithm MGD for detection.

## Modification

Add KD algorithm MGD for detection.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Review by wilxy — COMMENTED** [review]
Thank you for your excellent PR.
I made some comments, please contact me if you have any questions.

File: mmrazor/models/losses/mgd_loss.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        teacher_channels,
+        alpha_mgd=0.00002,
+        lambda_mgd=0.65,
+    ):
+        super(MGDLoss, self).__init__()
+        self.alpha_mgd = alpha_mgd
+        self.lambda_mgd = lambda_mgd
+
+        if student_channels != teacher_channels:
+            self.align = nn.Conv2d(
```
[the remark this answers is not in the document]
Comment (TinyTigerPan) [reply mmrazor/models/losses/mgd_loss.py:35]: Yes, I will change it later.

**Comment by TinyTigerPan** [comment]
1. Official accuracy can be reproduced in this commit. log and ckpt can be provided.
2. I will use `connector` to beauty my code.

**Comment by TinyTigerPan** [comment]
Finished! And I have trained 6 epochs, the loss and eval mAP looks right.

File: configs/distill/mmdet/mgd/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+| :------: | :-----: | :----------------------------------------------------------------------------------------------------------------------------: | :--------------------------------------------------------------------------------------------------------------------: | :-----: | :-: | :----: | :----: | :-------------------------------------------------: | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+|   FPN    |  COCO   | [RetinaNet-X101](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/retinanet/retinanet_x101-64x4d_fpn_1x_coco.py) | [RetinaNet-R50](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/retinanet/retinanet_r50_fpn_2x_coco.py) |   2x    |     |  41.0  |  37.4  | [config](mgd_fpn_retina_x101_retina_r50_2x_coco.py) | [teacher](https://download.openmmlab.com/mmdetection/v2.0/retinanet/retinanet_x101_64x4d_fpn_1x_coco/retinanet_x101_64x4d_fpn_1x_coco_20200130-366f5af1.pth) \|[model](<>) \| [log](<>) |
```
Comment (wilxy) [inline configs/distill/mmdet/mgd/README.md:19]: Please contact @pppppM to upload the relative `log` and `model` files.

File: mmrazor/models/architectures/connectors/mgd_connector.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import torch
+import torch.nn as nn
+
+from mmrazor.registry import MODELS
+from .base_connector import BaseConnector
+
+
+@MODELS.register_module()
+class MGDConncetor(BaseConnector):
```
Comment (wilxy) [inline mmrazor/models/architectures/connectors/mgd_connector.py:12]: Maybe `MGDConncetor` should be `MGDConnector`.
Add the test case for the `MGDConnector` in [test_connectors](https://github.com/open-mmlab/mmrazor/blob/dev-1.x/tests/test_models/test_architectures/test_connectors/test_connectors.py)

File: mmrazor/models/losses/mgd_loss.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+
+import torch
+import torch.nn as nn
+
+from mmrazor.registry import MODELS
+
+
+@MODELS.register_module()
+class MGDLoss(nn.Module):
```
Comment (wilxy) [inline mmrazor/models/losses/mgd_loss.py:10]: Add the test case of `MGDLoss` in [test_distillation_losses](https://github.com/open-mmlab/mmrazor/blob/dev-1.x/tests/test_models/test_losses/test_distillation_losses.py).
Add the docstring of `get_dis_loss`.

**Review by wilxy — COMMENTED** [review]
Good job!
It's ok to me after dealing with the following comments.

**Comment by TinyTigerPan** [comment]
Done! As for `log` and `model` files, since modified the code (use `connector`), I wondered if the ckpt will be changed, so the `ckpt` in frist version can not be applied to latest version. 
If MMRazor need the first version `ckpt` and `log`, I will provide. Moreover, if the code has no problem to merge, I can train it, but not now, maybe this weekend.

**Comment by TinyTigerPan** [comment]
Oops, some wrong with unitest, but I am not familiar with it, can you help me?

**Comment by wilxy** [comment]
> Oops, some wrong with unitest, but I am not familiar with it, can you help me?



> Oops, some wrong with unitest, but I am not familiar with it, can you help me?

File: tests/test_models/test_losses/test_distillation_losses.py [hunk 5]
```diff
@@ -204,3 +205,15 @@ def test_pkdloss(self):
...
+    def test_mgd_loss(self):
+        mgd_loss = MGDLoss(alpha_mgd=0.00002)
+        feats_S, feats_T = torch.rand(2, 256, 4, 4), torch.rand(2, 256, 4, 4)
+        loss = mgd_loss(feats_S, feats_T)
+        self.assertTrue(loss.numel() == 1)
+
+        num_stages = 4
+        feats_S = (torch.rand(2, 256, 4, 4) for _ in range(num_stages))
+        feats_T = (torch.rand(2, 256, 4, 4) for _ in range(num_stages))
+        loss = mgd_loss(feats_S, feats_T)
```
Comment (wilxy) [inline tests/test_models/test_losses/test_distillation_losses.py]: `feats_S` and `feats_T` are `tuple` of `torch.Tensor` here, while the `MGDLoss` only supports `torch.Tensor` as inputs.
You can change them to `torch.rand(2, 256, 4, 4)` to resolve the error in unittest.

Comment (wilxy) [reply tests/test_models/test_losses/test_distillation_losses.py]: There is also an error in [test_connectors](https://github.com/open-mmlab/mmrazor/actions/runs/3636395475/jobs/6136340778#step:13:165).

**Comment by TinyTigerPan** [comment]
Fixed, in torch 1.6, `torch.where` not support `Scalar`.

**Review by wilxy — APPROVED** [review]

File: configs/distill/mmdet/mgd/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+|   FPN    |  COCO   | [RetinaNet-X101](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/retinanet/retinanet_x101-64x4d_fpn_1x_coco.py) | [RetinaNet-R50](https://github.com/open-mmlab/mmdetection/blob/dev-3.x/configs/retinanet/retinanet_r50_fpn_2x_coco.py) |   2x    |     |  41.0  |  37.4  | [config](mgd_fpn_retina_x101_retina_r50_2x_coco.py) | [teacher](https://download.openmmlab.com/mmdetection/v2.0/retinanet/retinanet_x101_64x4d_fpn_1x_coco/retinanet_x101_64x4d_fpn_1x_coco_20200130-366f5af1.pth) \|[model](https://download.openmmlab.com/mmrazor/v1/mgd/mgd_fpn_retina_x101_retina_r50_2x_coco_20221209_191847-87141529.pth) \| [log](https://download.openmmlab.com/mmrazor/v1/mgd/mgd_fpn_retina_x101_retina_r50_2x_coco_20221209_191847-87141529.log) |
```
Comment (pppppM) [inline configs/distill/mmdet/mgd/README.md:19]: distilled mAP is empty

**Review by pppppM — APPROVED** [review]
